botwo/botwo.py

```diff
@@ -1,131 +1,144 @@
 
     def paginate(self, **kwargs):
-        paginator_to_call = self.paginator1
-        for map in self.mapping:
-            if fnmatch.fnmatch(kwargs["Bucket"], map["bucket_name"]):
-                if map.get("client"):
-                    if map.get("client") == self.client2_name:
-                        paginator_to_call = self.paginator2
-                    elif map.get("client") == self.client1_name:
-                        paginator_to_call = self.paginator1
-                    else:
-                        raise ValueError("Invalid client name " + map.get("client"))
-                else:
-                    raise ValueError("client isn't provided")
-
-                if map.get("mapped_bucket_name"):
-                    kwargs["Bucket"] = map.get("mapped_bucket_name")
-                if kwargs.get("Prefix"):
-                    if map.get("key_prefix"):
-                        kwargs["Prefix"] = map.get("keu_prefix") + kwargs["Prefix"]
-                break
-        return getattr(paginator_to_call, "paginate")(**kwargs)
-
-
-class Botwo(object):
+        for profile in self.config:
+            mapping = self.config[profile]
+
+            if fnmatch.fnmatch(kwargs["Bucket"], mapping.get("source_bucket_pattern")):
+                if kwargs["Prefix"]:
+                    if mapping.get("source_key_pattern"):
+                        if not fnmatch.fnmatch(kwargs["Prefix"], mapping.get("source_key_pattern")):
+                            continue
+                    if mapping.get("mapped_prefix"):
+                        kwargs["Prefix"] = mapping.get("mapped_prefix") + kwargs["Prefix"]
+                if mapping.get("mapped_bucket_name"):
+                    kwargs["Bucket"] = mapping.get("mapped_bucket_name")
+
+                return getattr(self.paginators.get(profile), "paginate")(**kwargs)
+        return getattr(self.default, "paginate")(**kwargs)
+
+
+class BotwoBuilder(object):
+    """ This class creates botwo client that wraps boto client:
+
+        * Holds boto clients and routes the requests between them by bucket/prefix configuration.
+        * Create its methods on the fly according to boto3 client AWS methods.
+        * Holds special treatment for functions that operate on multiple buckets or keys
+
+    """
+
     def __init__(self):
-        self.client1 = None
-        self.client2 = None
-        self.default_client = None
-        self.client1_name = None
-        self.client2_name = None
+        self.default = None
         self.mapping = None
-        self.meta = None
+        self.config = None
 
-    def create_client_class(self, client_mapping, config):
-        if type(client_mapping) is not dict:
-            raise TypeError("Invalid client mapping type: " + str(type(client_mapping)) + " expected dict")
-        if len(client_mapping) != 2:
-            raise TypeError("two clients is expected")
+    def build(self, mapping, config):
```

Comment:
This function needs a docstring according to PEP-8 and PEP-257.  That docstring should document `config` and `mapping`.

Reply:
Done